experimental/dds/tree2/src/core/tree/visitorUtils.ts

```diff
@@ -30,59 +30,115 @@ export function announceDelta(
 	deltaProcessor: { acquireVisitor: () => AnnouncedVisitor },
 	detachedFieldIndex: DetachedFieldIndex,
 ): void {
-	const visitor = announceVisitor(deltaProcessor.acquireVisitor());
+	const visitor = combineVisitors([deltaProcessor.acquireVisitor()]);
 	visitDelta(delta, visitor, detachedFieldIndex);
 	visitor.free();
 }
 
-export function combineVisitors(visitors: readonly DeltaVisitor[]): DeltaVisitor {
-	return {
-		free: () => visitors.forEach((v) => v.free()),
-		create: (...args) => visitors.forEach((v) => v.create(...args)),
-		destroy: (...args) => visitors.forEach((v) => v.destroy(...args)),
-		attach: (...args) => visitors.forEach((v) => v.attach(...args)),
-		detach: (...args) => visitors.forEach((v) => v.detach(...args)),
-		replace: (...args) => visitors.forEach((v) => v.replace(...args)),
-		enterNode: (...args) => visitors.forEach((v) => v.enterNode(...args)),
-		exitNode: (...args) => visitors.forEach((v) => v.exitNode(...args)),
-		enterField: (...args) => visitors.forEach((v) => v.enterField(...args)),
-		exitField: (...args) => visitors.forEach((v) => v.exitField(...args)),
-	};
+function isDeltaVisitor(obj: any): obj is DeltaVisitor {
+	return obj.create !== undefined;
+}
+
+function isAnnouncedVisitor(obj: any): obj is AnnouncedVisitor {
+	return obj.afterCreate !== undefined;
 }
 
-export function announceVisitor(visitor: AnnouncedVisitor): DeltaVisitor {
+export function combineVisitors(
+	visitors: readonly (DeltaVisitor | AnnouncedVisitor)[],
```

Comment:
I think it would be better to take in two separate lists instead of doing slightly dangerous runtime checks.